snuba/datasets/errors_replacer.py

```diff
@@ -0,0 +1,427 @@
+from snuba.datasets.schemas.tables import TableSchema, WritableTableSchema
+from snuba.processor import InvalidMessageType, _hashify
+from snuba.redis import redis_client
+from snuba.replacers.replacer_processor import (
+    Replacement,
+    ReplacementMessage,
+    ReplacerProcessor,
+)
+
+
+logger = logging.getLogger(__name__)
+
+EXCLUDE_GROUPS = object()
+NEEDS_FINAL = object()
+
+
+def get_project_exclude_groups_key(project_id: int) -> str:
+    return "project_exclude_groups:%s" % project_id
+
+
+def set_project_exclude_groups(project_id: int, group_ids: Sequence[int]) -> None:
+    """Add {group_id: now, ...} to the ZSET for each `group_id` to exclude,
+    remove outdated entries based on `settings.REPLACER_KEY_TTL`, and expire
+    the entire ZSET incase it's rarely touched."""
+
+    now = time.time()
+    key = get_project_exclude_groups_key(project_id)
+    p = redis_client.pipeline()
+
+    p.zadd(key, **{str(group_id): now for group_id in group_ids})
+    p.zremrangebyscore(key, -1, now - settings.REPLACER_KEY_TTL)
+    p.expire(key, int(settings.REPLACER_KEY_TTL))
+
+    p.execute()
+
+
+def get_project_needs_final_key(project_id: int) -> str:
+    return "project_needs_final:%s" % project_id
+
+
+def set_project_needs_final(project_id: int) -> Optional[bool]:
+    return redis_client.set(
+        get_project_needs_final_key(project_id), True, ex=settings.REPLACER_KEY_TTL
+    )
+
+
+def get_projects_query_flags(project_ids: Sequence[int]) -> Tuple[bool, Sequence[int]]:
+    """\
+    1. Fetch `needs_final` for each Project
+    2. Fetch groups to exclude for each Project
+    3. Trim groups to exclude ZSET for each Project
+
+    Returns (needs_final, group_ids_to_exclude)
+    """
+
+    s_project_ids = set(project_ids)
+    now = time.time()
+    p = redis_client.pipeline()
+
+    needs_final_keys = [
+        get_project_needs_final_key(project_id) for project_id in s_project_ids
+    ]
+    for needs_final_key in needs_final_keys:
+        p.get(needs_final_key)
+
+    exclude_groups_keys = [
+        get_project_exclude_groups_key(project_id) for project_id in s_project_ids
+    ]
+    for exclude_groups_key in exclude_groups_keys:
+        p.zremrangebyscore(
+            exclude_groups_key, float("-inf"), now - settings.REPLACER_KEY_TTL
+        )
+        p.zrevrangebyscore(
+            exclude_groups_key, float("inf"), now - settings.REPLACER_KEY_TTL
+        )
+
+    results = p.execute()
+
+    needs_final = any(results[: len(s_project_ids)])
+    exclude_groups = sorted(
+        {int(group_id) for group_id in sum(results[(len(s_project_ids) + 1) :: 2], [])}
+    )
+
+    return (needs_final, exclude_groups)
+
+
+class ErrorsReplacer(ReplacerProcessor):
+    def __init__(
+        self,
+        write_schema: WritableTableSchema,
+        read_schema: TableSchema,
+        required_columns: Sequence[str],
+        tag_column_map: Mapping[str, Mapping[str, str]],
+        promoted_tags: Mapping[str, Sequence[str]],
```

Comment:
This is the key: by splitting replacer processor, now the replacer does not need to fetch these parameters from the dataset since they are passed in the definition of the dataset itself.
Thus we do not have to deal with a brittle dataset interface that exposes elements that are supposed to be defined at the table storage level.